platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/IdpMetadata.java

```diff
@@ -119,66 +123,118 @@ private void initSingleSignOut() {
-          signingCertificate = certificate;
-        }
-
-        if (UsageType.ENCRYPTION.equals(key.getUse())) {
-          encryptionCertificate = certificate;
-        }
-
-        if (UsageType.SIGNING.equals(key.getUse())) {
-          signingCertificate = certificate;
-        }
-      }
+    if (descriptor == null) {
+      return;
     }
-  }
 
-  private EntityDescriptor getEntityDescriptor() {
-    Map<String, EntityDescriptor> edMap = entryDescriptions.get();
-    if (edMap == null) {
-      try {
-        edMap = parseMetadata();
-      } catch (IOException e) {
-        LOGGER.debug("Error parsing SSO metadata", e);
-        return null;
+    for (KeyDescriptor key : descriptor.getKeyDescriptors()) {
+      String certificate = null;
+      if (!key.getKeyInfo().getX509Datas().isEmpty()
+          && !key.getKeyInfo().getX509Datas().get(0).getX509Certificates().isEmpty()) {
+        certificate =
+            key.getKeyInfo().getX509Datas().get(0).getX509Certificates().get(0).getValue();
+      }
+
+      if (StringUtils.isBlank(certificate)) {
+        break;
       }
 
-      boolean updated = entryDescriptions.compareAndSet(null, edMap);
-      if (!updated) {
-        LOGGER.debug("Safe but concurrent update to serviceProviders map; using processed value");
+      if (UsageType.UNSPECIFIED.equals(key.getUse())) {
+        encryptionCertificate = certificate;
+        signingCertificate = certificate;
       }
+
+      if (UsageType.ENCRYPTION.equals(key.getUse())) {
+        encryptionCertificate = certificate;
+      }
+
+      if (UsageType.SIGNING.equals(key.getUse())) {
+        signingCertificate = certificate;
+      }
+    }
+  }
+
+  /**
+   * If the metadata is past its validity date, the SAML standard prohibits the metadata data from
+   * being used. The SAML entity is cleared if the metadata is invalid. This forces the class to
+   * attempt to retrieve the SAML entity from its source. If metadata is expired, it can continue to
+   * be used, but class attempts to get a new copy of it from the source. If successful, it clears
+   * the entity cache. If the entity cache is empty, the class attempts to retrieve it from the
+   * source.
+   *
+   * @return The root SAML entity descriptor or null
+   */
+  @VisibleForTesting
+  @Nullable
+  EntityDescriptor getEntityDescriptor() {
+
+    if (!isMetadataValid()) {
+      LOGGER.debug("SSO metadata is invalid. Purging metadata cache");
+      entityData.set(null);
     }
 
-    Set<Map.Entry<String, EntityDescriptor>> entries = edMap.entrySet();
-    if (!entries.isEmpty()) {
-      return entries.iterator().next().getValue();
+    EntityData ed =
+        Optional.ofNullable(entityData.get())
+            .orElse(
+                Optional.ofNullable(parseMetadata())
+                    .map(this::extractRootEntityFromMap)
+                    .map(EntityData::new)
+                    .orElse(null));
+
+    if (isMetadataExpired()) {
```

Comment:
❓ This should occur before the reparsing on lines 176-182, shouldn't it? Otherwise, we might come into this method with the `AtomicReference` set, find that is expired, but go ahead and reset it right back to the same value?